presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java

```diff
@@ -738,95 +777,164 @@ public void dropRole(String role)
         }
         finally {
-            tablePrivilegesCache.invalidate(new UserTableKey(grantee, databaseName, tableName));
+            invalidateTablePrivilegesCache(grantee, databaseName, tableName);
         }
     }
 
     @Override
-    public Set<HivePrivilegeInfo> listTablePrivileges(String databaseName, String tableName, PrestoPrincipal principal)
+    public Set<HivePrivilegeInfo> listTablePrivileges(MetastoreContext metastoreContext, String databaseName, String tableName, PrestoPrincipal principal)
+    {
+        return get(tablePrivilegesCache, new KeyAndContext<>(withContext(metastoreContext), new UserTableKey(principal, databaseName, tableName)));
+    }
+
+    public Set<HivePrivilegeInfo> loadTablePrivileges(KeyAndContext<UserTableKey> loadTablePrivilegesKey)
     {
-        return get(tablePrivilegesCache, new UserTableKey(principal, databaseName, tableName));
+        return delegate.listTablePrivileges(loadTablePrivilegesKey.getContext(), loadTablePrivilegesKey.getKey().getDatabase(), loadTablePrivilegesKey.getKey().getTable(), loadTablePrivilegesKey.getKey().getPrincipal());
+    }
+
+    private static class KeyAndContext<T>
+    {
+        private final MetastoreContext context;
+        private final T key;
+        private final int hashCode;
+
+        public KeyAndContext(MetastoreContext context, T key)
+        {
+            this.context = requireNonNull(context, "identity is null");
+            this.key = requireNonNull(key, "key is null");
+            this.hashCode = Objects.hash(context, key);
+        }
+
+        public MetastoreContext getContext()
+        {
+            return context;
+        }
+
+        public T getKey()
+        {
+            return key;
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            KeyAndContext<?> other = (KeyAndContext<?>) o;
+            return hashCode == other.hashCode &&
+                    Objects.equals(context, other.context) &&
+                    Objects.equals(key, other.key);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return hashCode;
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("context", context)
+                    .add("key", key)
+                    .toString();
+        }
     }
 
-    public Set<HivePrivilegeInfo> loadTablePrivileges(String databaseName, String tableName, PrestoPrincipal principal)
+    private MetastoreContext withContext(MetastoreContext context)
     {
-        return delegate.listTablePrivileges(databaseName, tableName, principal);
+        return isMetastoreImpersonationEnabled ? context : NO_IMPERSONATION_METASTORE_CONTEXT;
```

Comment:
Hi, right now the caching key is including this MetastoreContext, and this method is just trying to use a single NO_IMPERSONATION_METASTORE_CONTEXT if the impersonation is not enabled. 

I do see the problems of including the whole MetastoreContext into the caching key since we are actually only using the identity (username) for it. But it is not a concern now for this PR. But we can change this class to include username only.

Reply:
Basically what I'm trying to suggest is to instead of trying to cache a single context in a static variable let's replace only the username if the impersonation is disabled. It is preffered to prevent accidental mistakes in the future if more fields is added to the context. So basically what I'm suggesting is something like this:


```
private KeyAndContext<T> getCachingKey(MetastoreContext context, T key) {
  String cachingUsername = isMetastoreImpersonationEnabled ? context.getUsername() : "no-impersonation-caching-user";
  MetastoreContext cachingContext = new MetastoreContext(cachingUsername);
  return new KeyAndContext(cachingContext, key);
}
```